src/state.rs

```diff
@@ -0,0 +1,58 @@
+//! Program state accounts
+
+use {
+    solana_program::clock::UnixTimestamp, solana_program::pubkey::Pubkey,
+    spl_pod::optional_keys::OptionalNonZeroPubkey,
+};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum AccountType {
+    Uninitialized,
+    Config,
+    Stake,
+}
+
+/// Configuration for a staking system
+pub struct Config {
+    /// After a deactivation, defines the number of seconds that must pass before
+    /// the stake is inactive and able to be withdrawn
+    cooldown_time_secs: u64,
+    /// The maximum proportion that can be deactivated at once, given as basis
+    /// points (1 / 10,000)
+    max_deactivation_basis_points: u16,
+    /// Total number of tokens delegated to the system. Since anyone can transfer
+    /// tokens into the vault without passing through the program, this number
+    /// is maintained independently.
```

Comment:
There is still the transfer hook though, right? Maybe I'm not understanding, but we could still capture those transfers.

Maybe you mean this is specifically the tokens that have gone through the program and are effectively delegated/staked?

Reply:
In the governance program, to figure out the outcome of a vote, we need to know what proportion of stake has said "yes" or "no". The stake account stores the voting power of an individual staker, but we need to know how much voting power is part of the network.

To simplify things, we might think to simply read the amount of tokens in the PAL staker escrow. However, if we do that, someone could maliciously just transfer tokens directly into the escrow account to fudge the proportions.

For example, if there are 100 PAL tokens in the escrow, and you have a 51 PAL staked, you could immediately win any vote. But if I transfer 300 PAL tokens into the escrow without going through the staking procedure, then it would be impossible for the governance to ever resolve votes, since we could only get up to 100 PAL in voting power, and the system things there's 400 PAL in total voting power.

So to get around that, we store the total amount staked here

Reply:
Ok gotcha, thanks for the explanation!